### PR TITLE
[utils] Add driver for determining imagenet topk accuracy

### PR DESCRIPTION
**Pull request by jfix71**
*Description*: This script is used as a driver to calculate Top-1 and Top-5 accuracy across Imagenet. It will resize+center-crop images prior to inference. This resize+center-crop can also be done to pre-process the images without performing inference, so that the images can be used many times repeatedly without having to resize+center-crop them over and over again.

On Resnet50, here is what I found (showing results with and without specialized imagenet normalization https://github.com/pytorch/glow/pull/2007):

- Float: 
  - Interpreter/CPU Backend (same results):
    - Standard normalization:
      - Top-1 accuracy: 0.6846
      - Top-5 accuracy: 0.8863
    - Specialized Imagenet normalization:
      - Top-1 accuracy: 0.7613
      - Top-5 accuracy: 0.9293

- Quantized:
  - Interpreter:
    - Standard normalization:
      - Top-1 accuracy: 0.6682
      - Top-5 accuracy: 0.8767
    - Specialized Imagenet normalization:
      - Top-1 accuracy: 0.7500
      - Top-5 accuracy: 0.9236
  - CPU Backend:
    - Standard normalization:
      - Top-1 accuracy: 0.6731
      - Top-5 accuracy: 0.8795
    - Specialized Imagenet normalization:
      - Top-1 accuracy: 0.7497
      - Top-5 accuracy: 0.9236

*Testing*: Tested it with many different inputs.

*Documentation*: Added to `Testing.md`

### CODE REVIEW

**Comment by jfix71** [comment]
I'm by no means a python expert, so happy to get suggestions here!

File: utils/imagenet_topk_accuracy_driver.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+## top1_count, and @param top5_count.
+def print_topk_accuracy(total_image_count, top1_count, top5_count):
+    top1_accuracy = float(top1_count) / float(total_image_count)
+    top5_accuracy = float(top5_count) / float(total_image_count)
+    print("\tTop-1 accuracy: " + "{0:.4f}".format(top1_accuracy))
+    print("\tTop-5 accuracy: " + "{0:.4f}".format(top5_accuracy))
+
+
+def main():
+    global args
```
Comment (bertmaher) [inline utils/imagenet_topk_accuracy_driver.py]: Can you avoid making `args` global, and pass the relevant bits down instead?

**Comment by rdzhabarov** [comment]
This is amazing! Where do you get inputs (expected outputs) for the accuracy test?

File: utils/imagenet_topk_accuracy_driver.py (hunk 1 above)
Comment (jfix71) [reply utils/imagenet_topk_accuracy_driver.py]: Done!

**Comment by jfix71** [comment]
@rdzhabarov External users can download labeled images from the [Imagenet dataset](http://www.image-net.org/)

File: utils/imagenet_topk_accuracy_driver.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+import tempfile
+
+import PIL.Image as Image
+
+import torchvision.transforms as transforms
+
+parser = argparse.ArgumentParser(description='Glow image-classifier Driver for TopK ImageNet Calculation')
+
+parser.add_argument('--validation-images-dir', metavar='DIR', required=True,
+                    help=" Path to the directory containing the validation set "
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: nit: " Path -> "Path

File: utils/imagenet_topk_accuracy_driver.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+def accimage_loader(path):
+    import accimage
+    try:
+        return accimage.Image(path)
+    except IOError:
+        # Potentially a decoding problem, fall back to PIL.Image
+        return pil_loader(path)
+
+def default_image_loader(path):
+    from torchvision import get_image_backend
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: please don't do import-not-at-top

File: utils/imagenet_topk_accuracy_driver.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    ])
+
+    resized_img = transform_resize(img)
+    resized_img.save(path_to_new_img, format="png")
+
+## Used to pre-process an input set of images. Takes a string of a directory
+# @param validation_images_dir and saves the cropped subset of the images in a
+# subdirectory `processed/`, which must not yet exist.
+def save_centered_cropped_dataset(validation_images_dir):
+    assert os.path.exists(validation_images_dir), "Validation directory does not exist: " + validation_images_dir
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: This and some other lines are longer than 80 chars. Could you use pylint or something like this?

File: utils/imagenet_topk_accuracy_driver.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+
+    try:
+        p = pexpect.spawn(args.image_classifier_cmd, logfile=sys.stdout,
+                          timeout=None)
+
+        # Create a temporary directory to store the transformed image we
+        # classify (if applicable) and the log of image-classifer output.
+        tmp_dir_name = tempfile.mkdtemp()
+        path_to_tmp_log = os.path.join(tmp_dir_name, "log.txt")
+        fout = file(path_to_tmp_log,'w')
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: space between args please

File: utils/imagenet_topk_accuracy_driver.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+                top5_labels = []
+                for x in xrange(5):
+                    label_and_prob = p.readline()
+                    # Get the label from the line.
+                    label = label_and_prob.split()[1]
+                    top5_labels.append(int(label))
+
+                expected_label = img_labels[img_index + batch_idx]
+                if expected_label == top5_labels[0]:
+                    top1_count = top1_count + 1
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: you can do `+= 1`

File: utils/imagenet_topk_accuracy_driver.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+    assert total_image_count % args.batch_size == 0, "Total number of images must be divisible by batch size"
+
+    if args.verbose:
+        print("Running image classifier with: " + args.image_classifier_cmd)
+
+    tmp_dir_name = ""
+
+    try:
+        p = pexpect.spawn(args.image_classifier_cmd, logfile=sys.stdout,
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: can you move this line after `fout = ` and pass `logfile=fout`?

File: utils/imagenet_topk_accuracy_driver.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+            if os.path.isfile(orig_img_path):
+                processed_img_path = os.path.join(processed_dir_path,
+                                                  orig_img_filename)
+                resize_and_save_image(orig_img_path, processed_img_path)
+
+## @returns a list of strings (of length equal to the batch size) which are
+# paths to images to do inference on. @param img_paths is the set of all image
+# paths, @param img_index is the next index to use in @param img_paths, and
+# @param tmp_dir_name is the location of where to save the images if necessary.
+def get_curr_img_paths(img_paths, img_index, tmp_dir_name, batch_size,
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: I see there is a lot of duplicated logic between the function above and this one. Could you do "crop and center" for all images together before starting to run image-classifier? That way, all this logic can go into one function. It can even cache automatically by looking at whether "processed" dir already exists.

File: utils/imagenet_topk_accuracy_driver.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+                      print("")
+
+    finally:
+        p.close(force=True)
+
+        # Remove the temp directory we used to save the images and log.
+        if tmp_dir_name:
+            shutil.rmtree(tmp_dir_name)
+
+    print("\nCompleted running; Final Top-1/5 accuracy across " +
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: I see a mix of double-quotes and single quotes for strings. And also `print` sometimes with brackets and sometimes without. I believe pylint can detect this too.

File: utils/imagenet_topk_accuracy_driver.py (hunk 3 above)
Comment (jfix71) [reply utils/imagenet_topk_accuracy_driver.py]: I removed this one, however the accimage above is still there -- it seems like a conditional import based on the if below (I took this code snippet from a different run script). I'm not sure of a good way to get around this, do you have any suggestion?

Comment (artemrakhov) [reply utils/imagenet_topk_accuracy_driver.py]: For accimage it's actually fine: it may not be present.

File: utils/imagenet_topk_accuracy_driver.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+        curr_img_paths.append(img_path)
+
+    return curr_img_paths
+
+## Verifies that the @param image_classifier_cmd is well formatted via
+# assertions.
+def verify_spawn_cmd(image_classifier_cmd):
+    split_cmd = image_classifier_cmd.split()
+
+    assert "image-classifier" in split_cmd[0], \
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: Using `\` for line breaks is discouraged by many style guides. You can use:
```
assert "image-classifier" in split_cmd[0], (
    "image-classifier binary must be provided first.")
```

File: utils/imagenet_topk_accuracy_driver.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+        torchvision.transforms.CenterCrop(224),
+    ])
+
+    resized_img = transform_resize(img)
+    resized_img.save(path_to_new_img, format="png")
+
+## Used to pre-process an input set of images. Takes a string of a directory
+# @param validation_images_dir and saves the cropped subset of the images in a
+# subdirectory `processed/`, which must not yet exist.
+def save_centered_cropped_dataset(validation_images_dir):
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: I think you can use this function for both "only-resize-and-save" and "resize-input-images" cases. Add a new argument "target_dir", and pass either `join(validation_images_dir, "processed")` or tmpdir (only run it if "resize-input-images"). 

So in case of "resize-input-images" you'll do all resizing in advance before starting classification. You'll reassign `img_paths` variable, and instead of calling `get_curr_img_paths` you'll just do: `img_paths[img_index, img_index + batch_size]`. Cleanup will happen automatically, when you remove tmpdir after `finally:`.

File: utils/imagenet_topk_accuracy_driver.py (hunk 10 above)
Comment (jfix71) [reply utils/imagenet_topk_accuracy_driver.py]: Done

File: utils/imagenet_topk_accuracy_driver.py (hunk 11 above)
Comment (jfix71) [reply utils/imagenet_topk_accuracy_driver.py]: >So in case of "resize-input-images" you'll do all resizing in advance before starting classification. 

Got it -- yeah that could work, but the downside here is the resized images must all exist at once. So for example the 50k images I was looking at would take up 4.2GB temporarily. I felt it was a waste to create that much temporary space on disk, but if you prefer that version in order to remove some duplicated code I can change it.

Comment (artemrakhov) [reply utils/imagenet_topk_accuracy_driver.py]: Agree, that's a lot of space. But my understanding is that in current version we only clean up once, in the very end.

Comment (jfix71) [reply utils/imagenet_topk_accuracy_driver.py]: Ah, actually the current version just reuses the same image name over and over again at [this line](https://github.com/pytorch/glow/pull/2009/files#diff-4ea927ddc0aaa413fc99b55398567f7aR182). So there are only ever `batch_size` temp images in the temp directory which is then cleaned up in the end. I can clarify that in the comments there if we proceed with this version.

Comment (artemrakhov) [reply utils/imagenet_topk_accuracy_driver.py]: Oh, I totally missed this. Well, feel free to leave this logic then.

File: utils/imagenet_topk_accuracy_driver.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+
+                expected_label = img_labels[img_index + batch_idx]
+                if expected_label == top5_labels[0]:
+                    top1_count += 1
+                if expected_label in top5_labels:
+                    top5_count += 1
+
+            curr_completed_count = img_index + batch_size
+            if curr_completed_count % 100 == 0:
+                print "Finished image index %s out of %s" % (
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: can you use %d and not cast to `str`?

File: utils/imagenet_topk_accuracy_driver.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+                else:
+                    print ""
+
+    finally:
+        classifier_proc.close(force=True)
+
+        # Remove the temp directory we used to save the images and log.
+        shutil.rmtree(tmp_dir_name)
+
+    print "\nCompleted running; Final Top-1/5 accuracy across %s images:" % (
```
Comment (artemrakhov) [inline utils/imagenet_topk_accuracy_driver.py]: can you use %d and not cast to `str`?

**Comment by jfix71** [comment]
Thanks for the reviews @artemrakhov!
